### ISSUE TITLE
On workflow module optimisation

### ISSUE THREAD

**Issue by hyzx86** [OrchardCMS/OrchardCore#16107]
### Is your feature request related to a problem? Please describe.
 
To date, we have collected a number of questions and suggestions for the workflow module
Below I've summarised and collated these encounters, and we've tried to consolidate them into one PR as much as possible, as separating them into different PRs may require multiple changes to the multi-table structure, and may also require the introduction of many destructive updates

#### Workflow support for versioning, including audit information

- Upgrade workflows diagramming library to a modern digramming ui library #16067
- Implement workflow versioning and keep the existing instance upon import #15496
- Workflows need audit trail support #11608

#### Other enhancements

Workflow Type
- Add title templates to the workflow, like `TitlePart` does 
- Add a `Description` field to a workflow type，can save text that describes the process.
- Add a `DisplayName` field for a short description to distinguish it from the technical name.

Workflow Instance
- Add workflow instance title editor for workflow types
- Adding display names to workflow types
- Add `ExecutedOnUtc` field to workflow instances, which can be used for sorting
- Adding audit information to workflow types, `CreatedUtc`,`ModifiedUtc`,`ModifiedBy`,`CreatedBy`

If there are any other suggestions, please add them.

### Describe the solution you'd like

I am currently trying to implement these changes in PR https://github.com/OrchardCMS/OrchardCore/pull/16043

**Comment by giannik** [comment]
@hyzx86 also a description field for the workflow type definition would be usefull (along with display name).

**Comment by hyzx86** [comment]
> @hyzx86 also a description field for the workflow type definition would be usefull (along with display name). 

Yes, I agree, it could be used to keep some change logs or business notes or something like that.

**Comment by giannik** [comment]
> > @hyzx86 also a description field for the workflow type definition would be usefull (along with display name).
> 
> Yes, I agree, it could be used to keep some change logs or business notes or something like that.

Basically from list of workflows  to see  and remember  what it does and the outcomes  so you dont have to look into the designer  to recall what its doing.

**Comment by hyzx86** [comment]
> > > @hyzx86 also a description field for the workflow type definition would be usefull (along with display name).此外，工作流类型定义的描述字段（以及显示名称）也很有用。
> > 
> > 
> > Yes, I agree, it could be used to keep some change logs or business notes or something like that.是的，我同意，它可以用来保存一些更改日志或业务笔记或类似的东西。
> 
> Basically from list of workflows to see and remember what it does and the outcomes so you dont have to look into the designer to recall what its doing.基本上从工作流程列表中查看并记住它的作用和结果，因此您不必查看设计师来回忆它的作用。

That would be DisplayName. 😁

**Comment by giannik** [comment]
the display name would : 
Display name :  Front end Contact form submission
The desription would be : 

handles submissions from form http://mysite.com. Sends emails to moderator and  integrates with crm bla bla....

Display name is something short and description expands more.

**Comment by hyzx86** [comment]
Hi @Piedone ，In addition to using x6 as a process designer, we may need a `json` driven form designer for dynamic form generation, I would recommend [amis](https://baidu.github.io/amis/zh-CN/components/crud) Do we have any limitations or rules regarding the introduction of external packages?

**Comment by giannik** [comment]
@hyzx86 personally i would not change anything else apart from  replacing js plumb with x6 . changing the activity  UI forms to depend on a javascript library is not something i would prefer.

**Comment by hyzx86** [comment]
> @hyzx86 personally i would not change anything else apart from replacing js plumb with x6 . changing the activity UI forms to depend on a javascript library is not something i would prefer. 

This is not that simple. First of all, our current workflow node editing is based on form submissions, and we want to refactor it to SPA, so we first need to change the way each node is saved from a per-submission to the server to a client-side staging.
In addition, I'd like to enhance the workflow node editing to support multiple methods, like elsa- where a single configuration item supports raw text, liquid expressions, and js expressions.
My plan is to abstract the process node definitions in c# code, generate a json structure, and then have the front-end components build the required UI. 
We can use amis or any other front-end component to generate the edit form.

**Comment by giannik** [comment]
......., and we want to refactor it to SPA, .....

why have you come to a conclusion that we need to refactor it to a SPA ?

**Comment by hyzx86** [comment]
> why have you come to a conclusion that we need to refactor it to a SPA ? 

This is a change based on workflow versioning, as currently every time a workflow node is saved the entire workflow type is directly submitted and saved to the database

**Comment by giannik** [comment]
and why is that a problem ? just save the entire workflow json and do a diff evalutaion of the changes . content item versioning works the same.

**Comment by hyzx86** [comment]
A version will be generated for each node save, are you sure this is appropriate?

**Comment by Piedone** [comment]
> Hi @Piedone ，In addition to using x6 as a process designer, we may need a `json` driven form designer for dynamic form generation, I would recommend [amis](https://baidu.github.io/amis/zh-CN/components/crud). Do we have any limitations or rules regarding the introduction of external packages?

I don't think we have anything formalized, but I'd say they need to have a license compatible with OC's, actively developed, have a sufficiently high quality, and with everything available in English (as well).

**Comment by weirdyang** [comment]
Will the audit trail eventually be added for workflows?

**Comment by Piedone** [comment]
It's possible that we'll work on it in the foreseeable future, yes, but not certain.

**Comment by weirdyang** [comment]
Thanks, I'm currently working on one following the content structure, but saw this PR and was wondering if it would cause any conflicts.

Is it recommended to create a separate audit trail table to avoid any future conflicts?

**Comment by Piedone** [comment]
Do you mean you're working on Audit Trail support for content definition changes? That would be great! For that, you don't need to create a separate Audit Trail table, if you mean a SQL table.

**Comment by weirdyang** [comment]
No I'm trying to implement audit trail for workflows using the same structure for content items,
Apologies for any confusion

**Comment by Piedone** [comment]
You can do that, I'd perhaps start with that too. What do you mean by "create a separate audit trail table to avoid any future conflicts", though? If you mean the table in HTML, then you'd need to add some kind of new table there, yes, but the basic Audit Trail record data (like who did it and when) should still come from the common shape.
